traits/trait_dict_object.py

```diff
@@ -57,18 +50,314 @@ def __init__(self, added=None, changed=None, removed=None):
+            self.key_validator = key_validator
+
+        if value_validator is not None:
+            self.value_validator = value_validator
+
+        if notifiers is None:
+            notifiers = []
+        self.notifiers = notifiers
+
+        if value is None:
+            value = {}
+
+        items = value.items() if isinstance(value, dict) else value
+        value = {self.key_validator(key): self.value_validator(value)
+                 for key, value in items}
+
+        super().__init__(value)
+
+    def notify(self, removed, added, changed):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_dict, removed, added, changed)
+
+        Any return values are ignored.
+        """
+
+        for notifier in self.notifiers:
+            notifier(self, removed=removed, added=added, changed=changed)
+
+    # -- dict interface -------------------------------------------------------
+
+    def __setitem__(self, key, value):
+        """ Set a value for the key, implements self[key] = value.
+
+        Parameters
+        ----------
+        key : A hashable object.
+            The key for the value.
+        value : any
+            The value to set for the corresponding key.
+        """
+
+        removed = {}
+        validated_key = self.key_validator(key)
+        validated_value = self.value_validator(value)
+
+        if validated_key in self:
+            changed = {validated_key: self[validated_key]}
+            added = {}
+        else:
+            changed = {}
+            added = {validated_key: validated_value}
+
+        super().__setitem__(validated_key, validated_value)
+        self.notify(removed=removed, added=added, changed=changed)
+
+    def __delitem__(self, key):
+        """ Delete the item from the dict indicated by the key.
+
+        Parameters
+        ----------
+        key : A hashable object.
+            The key to be deleted.
+
+        Raises
+        ------
+        KeyError
+            If the key is not found.
+        """
+
+        removed = {key: self[key]} if key in self else {}
+        super().__delitem__(key)
+        self.notify(removed=removed, added={}, changed={})
+
+    def clear(self):
+        """ Remove all items from the dict. """
+        was_empty = (self == {})
+        removed = self.copy()
+
+        super().clear()
+        if not was_empty:
+            self.notify(removed=removed, added={}, changed={})
+
+    def update(self, other):
+        """ Update the values in the dict by the new dict or an iterable of
+        key-value pairs.
+
+        Parameters
+        ----------
+        other : dict or iterable
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        items = other.items() if isinstance(other, dict) else other
```

Comment:
As with the `__init__`, let's use the presence of the `keys` attribute when deciding whether to treat as a mapping or an iterable of pairs.